planetiler-core/src/main/java/com/onthegomap/planetiler/stream/WriteableProtoStreamArchive.java

```diff
@@ -0,0 +1,184 @@
+    if (envelope == null) {
+      return null;
+    }
+    return StreamArchiveProto.Envelope.newBuilder()
+      .setMinX(envelope.getMinX())
+      .setMaxX(envelope.getMaxX())
+      .setMinY(envelope.getMinY())
+      .setMaxY(envelope.getMaxY())
+      .build();
+  }
+
+  private static StreamArchiveProto.CoordinateXY toExportData(CoordinateXY coord) {
+    if (coord == null) {
+      return null;
+    }
+    return StreamArchiveProto.CoordinateXY.newBuilder()
+      .setX(coord.getX())
+      .setY(coord.getY())
+      .build();
+  }
+
+  private static StreamArchiveProto.VectorLayer toExportData(VectorLayer vectorLayer) {
+    final var builder = StreamArchiveProto.VectorLayer.newBuilder();
+    builder.setId(vectorLayer.id());
+    vectorLayer.fields().entrySet().forEach(e -> {
+      var exportType = switch (e.getValue()) {
+        case NUMBER -> StreamArchiveProto.VectorLayer.FieldType.FIELD_TYPE_NUMBER;
+        case BOOLEAN -> StreamArchiveProto.VectorLayer.FieldType.FIELD_TYPE_BOOLEAN;
+        case STRING -> StreamArchiveProto.VectorLayer.FieldType.FIELD_TYPE_STRING;
+      };
+      builder.putFields(e.getKey(), exportType);
+    });
+    vectorLayer.description().ifPresent(builder::setDescription);
+    vectorLayer.minzoom().ifPresent(builder::setMinZoom);
+    vectorLayer.maxzoom().ifPresent(builder::setMaxZoom);
+    return builder.build();
+  }
+
+  private static <T> void setIfNotNull(Consumer<T> setter, T value) {
+    if (value != null) {
+      setter.accept(value);
+    }
+  }
+
+  private static class ProtoTileArchiveWriter implements TileWriter {
+
+    private final OutputStream out;
+
+    ProtoTileArchiveWriter(OutputStream out) {
+      this.out = out;
+    }
+
+    @Override
+    public void write(TileEncodingResult encodingResult) {
+      final TileCoord coord = encodingResult.coord();
+      final StreamArchiveProto.TileEntry tile = StreamArchiveProto.TileEntry.newBuilder()
+        .setZ(coord.z())
+        .setX(coord.x())
+        .setY(coord.y())
+        .setEncodedData(ByteString.copyFrom(encodingResult.tileData()))
+        .build();
+
+      final StreamArchiveProto.Entry entry = StreamArchiveProto.Entry.newBuilder()
+        .setTile(tile)
+        .build();
+
+      try {
+        entry.writeDelimitedTo(out);
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+
+    @Override
+    public void close() {
+      try {
+        out.close();
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+  }
+
+  // for quick testing
+  // note: do not use nio (Files.newInputStream) for pipes
+  public static void main(String[] args) throws IOException {
+    try (var in = new FileInputStream(args[0])) {
+      StreamArchiveProto.Entry entry;
+      while ((entry = StreamArchiveProto.Entry.parseDelimitedFrom(in)) != null) {
+        System.out.println(entry);
+      }
+    }
+  }
```

Comment:
Remove this testing code?

Reply:
I tend to leave such code for testing if it's simple and handy, but okay, I'll remove